### PR TITLE
sunxi: add support for xr819 wifi on OrangePi Zero

### PR DESCRIPTION
**Pull request by ysc3839**
Based on @melsem 's work #2026 #2016 #2013 #2012.

Build test: passed.
Run test:
WPA2-PSK Force CCMP (AES)
iperf3 speed test using my mobile phone:
```
$ iperf3 -c 192.168.2.1
Connecting to host 192.168.2.1, port 5201
[  4] local 192.168.2.146 port 59470 connected to 192.168.2.1 port 5201
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-1.00   sec  3.53 MBytes  29.6 Mbits/sec                  
[  4]   1.00-2.04   sec  0.00 Bytes  0.00 bits/sec                  
[  4]   2.04-3.02   sec  1.25 MBytes  10.7 Mbits/sec                  
[  4]   3.02-4.02   sec  1.25 MBytes  10.5 Mbits/sec                  
[  4]   4.02-5.04   sec  0.00 Bytes  0.00 bits/sec                  
[  4]   5.04-6.03   sec  1.25 MBytes  10.5 Mbits/sec                  
[  4]   6.03-7.03   sec  1.25 MBytes  10.5 Mbits/sec                  
[  4]   7.03-8.03   sec  0.00 Bytes  0.00 bits/sec                  
[  4]   8.03-9.00   sec  1.25 MBytes  10.8 Mbits/sec                  
[  4]   9.00-10.02  sec  0.00 Bytes  0.00 bits/sec                  
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-10.02  sec  9.78 MBytes  8.19 Mbits/sec                  sender
[  4]   0.00-10.02  sec  9.11 MBytes  7.62 Mbits/sec                  receiver

iperf Done.
```
```
$ iperf3 -c 192.168.2.1 -R
Connecting to host 192.168.2.1, port 5201
Reverse mode, remote host 192.168.2.1 is sending
[  4] local 192.168.2.146 port 59510 connected to 192.168.2.1 port 5201
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-1.00   sec   803 KBytes  6.58 Mbits/sec                  
[  4]   1.00-2.01   sec   789 KBytes  6.42 Mbits/sec                  
[  4]   2.01-3.00   sec   841 KBytes  6.94 Mbits/sec                  
[  4]   3.00-4.00   sec   899 KBytes  7.37 Mbits/sec                  
[  4]   4.00-5.04   sec  1.06 MBytes  8.51 Mbits/sec                  
[  4]   5.04-6.00   sec   617 KBytes  5.26 Mbits/sec                  
[  4]   6.00-7.00   sec   800 KBytes  6.57 Mbits/sec                  
[  4]   7.00-8.00   sec   730 KBytes  5.96 Mbits/sec                  
[  4]   8.00-9.01   sec   933 KBytes  7.63 Mbits/sec                  
[  4]   9.01-10.00  sec   673 KBytes  5.54 Mbits/sec                  
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bandwidth       Retr
[  4]   0.00-10.00  sec  9.10 MBytes  7.63 Mbits/sec    0             sender
[  4]   0.00-10.00  sec  8.23 MBytes  6.90 Mbits/sec                  receiver

iperf Done.
```

### CODE REVIEW

**Comment by ynezz** [comment]
First of all you, I recommend you re-reading https://openwrt.org/submitting-patches one more time.

Honestly, I'm not sure, if we want to include driver/firmware of [such quality in the tree](https://github.com/fifteenhex/xradio#read-this):

> This driver just about works. If it loads and starts working normally it usually doesn't have any major issues until you try to reboot. The structure of the driver isn't great and it can lock up the kernel if it gets confused. It's not production ready for your large scale commercial IoT deployment.

> Also: The xr819 chip/firmware drops tons and tons of frames with FCS errors and this makes performance horrible at best. Most people have lost interest in having anything to do with the xr819 because of people being idiots and demanding that issues that are incredibly hard to fix without documentation be fixed because they spent $8 on a board and somehow people that got exactly zero of their $8 are responsible.

> Moral of the story: If you're going to post nasty things on the interwebs and demand people fix stuff because reasons at least have a bunch of packet dumps etc and have some idea about what you're talking about.

and [from](https://github.com/fifteenhex/xradio#issues):

> The firmware running on the xr819 is very crash happy and the driver is a bit stupid. For example the driver can get confused about how many packets of data the xr819 has for it to read and can try to read too many. The firmware on the xr819 responds by triggering an assert and shutting down. The driver gets a packet that tells it that the firmware is dead and shuts down the thread used to push and pull data but the rest of the driver and the os has no idea and if the os tries to interact with the driver everything starts to lock up.

> Pings from the device to the network are faster than from the network to the device. This seems to be because of latency between the interrupt and servicing RX reports from the device.

Anyway, putting apart this general quality and usability concerns, you shouldn't be adding any driver sources to the kernel package, but  probably use https://github.com/fifteenhex/xradio to get the driver sources instead. You should do the same with the firmware blob as well. Then you should avoid adding that kernel module to the default firmware image for OrangePi, as this would mean, that we would need to support this driver and I can't imagine, that anyone would be willing to do so. In other words, it might be probably acceptable to have it available as manually installable kernel module, but it's not acceptable to provide it as default kernel module as the users would expect, that it's tested and working combination, which nobody is going to guarantee.

Once you solve this issues with this PR, you would need to convince someone to actually merge it and due to that quality/usability concerns stated above, I doubt, that you could find someone.

So to sum it up, your best bet is to try upstream (or find someone) the driver (adding it to cw1200 driver?) and then eventually backport it back to OpenWrt.
